### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by eldritchconundrum**
not for merging

### CODE REVIEW

File: src/ast.fs [hunk 1]
```diff
@@ -2,10 +2,12 @@
 
 open Options.Globals
 
-type Ident = string
+type Ident = { name: string }
+let makeIdent name = { Ident.name = name }
```
Comment (laurentlb) [inline src/ast.fs:6]:
```
type Ident(name) =
    member this.Name = name
```

hasInlinePrefix and isTemporaryId should be methods

File: src/ast.fs [hunk 2]
```diff
@@ -84,8 +86,8 @@ let makeFunctionType ty name args sem =
 // values). We need to provide accessors for the developer (e.g. create macros for C/C++).
 type ExportedName = {
     ty: string  // "F" for hlsl functions, empty for vars
-    name: string
-    newName: string
+    name: Ident
+    newName: Ident
```
Comment (laurentlb) [inline src/ast.fs:90]: I suggest to keep a string here for now. It should make the PR simpler.

File: src/renamer.fs [hunk 3]
```diff
@@ -24,7 +24,7 @@ module private RenamerImpl =
         // Whether multiple functions can have the same name (but different arity).
```
Comment (laurentlb) [inline src/renamer.fs:24]: We should use Ident for AST nodes. `availableNames` should be a `string list`.
Also, maybe:
```
        varRenames: Map<string, Ident>
        funRenames: Map<string, Map<int, Ident>>
```

Comment (eldritchconundrum) [reply src/renamer.fs:24]: I'm planning Equals+GetHashCode for Ident

Comment (laurentlb) [reply src/renamer.fs:24]: Maps don't need hashable (it's a tree).
I don't think we want to compare Idents.

**Pull request closed without merging**